src/main/java/org/veupathdb/service/eda/ds/plugin/AbstractPlugin.java

```diff
@@ -646,22 +648,25 @@ public boolean validateImputeZeroesRequest(Map<String, DynamicDataSpecImpl> data
     return true;
   }
 
-  public String getRStudyVocabsAsString(DynamicDataSpecImpl dataSpec, ResponseFuture studyVocab) {
+  public String getRStudyVocabsAsString(DynamicDataSpecImpl dataSpec, ResponseFuture studyVocab) throws Exception {
     PluginUtil util = getUtil();
-    
-    // this assuming the first ancestor is the root one is a bit hacky, but i did the same in R so...
-    // obviously doing the same awful thing twice makes it ok. but doing any better is higher cost than i have time for right now :(
-    String readStudyVocabInR = "data.table::fread(" + studyVocab + ", header = FALSE)";
-    String studyVocabAsRTibble = "dplyr::reframe(dplyr::group_by(" + readStudyVocabInR + ", V1), " +
-                                         "values=paste0('veupathUtils::StudySpecificVocabulary(variable=veupathUtils::VariableSpec(entityId=" + getDynamicDataSpecEntityId(dataSpec)+ ", " + 
-                                                                                              "variableId=" + getDynamicDataSpecId(dataSpec) + ")" + 
-                                                          "vocabulary=c(',paste(V2, collapse=','),')," +
-                                                          "study=\"',V1,'\"'" +
-                                                          "studyIdColumnName='" + util.getEntityAncestorsAsRVectorString(getDynamicDataSpecEntityId(dataSpec), _referenceMetadata) + "[1])))";
-  
-    String studyVocabsListAsRString = "veupathUtils::StudySpecificVocabulariesByVariable(S4Vectors::SimpleList(paste(" + studyVocabAsRTibble + "[2], collapse=',')))";
-    
-    return studyVocabsListAsRString;
+    ByteArrayOutputStream out = new ByteArrayOutputStream();
+    useRConnectionWithRemoteFiles(Resources.RSERVE_URL, Map.of("studyVocab", studyVocab.getInputStream()), connection -> {
```

Comment:
I was intending on calling this helper from the same part of the associated plugin, so that we know we're working in the same r process for the subsequent call to plot.data... I'm wondering if we can pass that connection through to here as a method arg somehow?

Reply:
Yeah it's also less overhead to re-use the connection. So would you just want this method to take an R connection as an additional param and assign the vocab data structure to a variable within the R connection?